timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -52,46 +52,52 @@ void __global__ k_check_rebuild_coords_and_box_gather(
     }
 }
 
-template <typename RealType>
+template <typename RealType, int COORDS_DIM, int PARAMS_DIM>
 void __global__ k_gather_coords_and_params(
     const int N,
     const unsigned int *__restrict__ idxs,
     const RealType *__restrict__ coords,
     const RealType *__restrict__ params,
     RealType *__restrict__ gathered_coords,
     RealType *__restrict__ gathered_params) {
-
+    static_assert(COORDS_DIM == 3);
+    static_assert(PARAMS_DIM == PARAMS_PER_ATOM);
     int idx = blockIdx.x * blockDim.x + threadIdx.x;
-    int stride = gridDim.y;
-    int stride_idx = blockIdx.y;
-
     if (idx >= N) {
         return;
     }
 
+    const unsigned int atom_idx = idxs[idx];
+
     // Coords have 3 dimensions, params have 4
-    if (stride_idx < 3) {
-        gathered_coords[idx * 3 + stride_idx] = coords[idxs[idx] * 3 + stride_idx];
+#pragma unroll COORDS_DIM
+    for (int i = 0; i < COORDS_DIM; i++) {
+        gathered_coords[idx * COORDS_DIM + i] = coords[atom_idx * COORDS_DIM + i];
+    }
+#pragma unroll PARAMS_DIM
+    for (int i = 0; i < PARAMS_DIM; i++) {
+        gathered_params[idx * PARAMS_DIM + i] = params[atom_idx * PARAMS_DIM + i];
     }
-    gathered_params[idx * stride + stride_idx] = params[idxs[idx] * stride + stride_idx];
 }
 
-template <typename T>
+template <typename T, int D>
 void __global__ k_scatter_accum(
     const int N,
     const unsigned int *__restrict__ unique_idxs, // NOTE: race condition possible if there are repeated indices
     const T *__restrict__ gathered_array,
     T *__restrict__ array) {
-
+    static_assert(D <= PARAMS_PER_ATOM);
     int idx = blockIdx.x * blockDim.x + threadIdx.x;
-    int stride = gridDim.y;
-    int stride_idx = blockIdx.y;
 
     if (idx >= N) {
         return;
     }
+    const unsigned int atom_idx = unique_idxs[idx];
```

Comment:
nit: would prefer to keep the naming generic here; maybe `dest_idx` or similar?

Reply:
Renamed in https://github.com/proteneer/timemachine/pull/1124/commits/0cdec51609225a1e4f6afc4456b2c354de5c5a5f